### PR TITLE
Apply some updates from my personal fork to the new baseline.

### PR DESCRIPTION
**Pull request by jamie-zaharopoulos-fh**
**Summary**

This PR fixes some issues and improvements from my personal fork

This PR fixes/implements the following **bugs/features**

* [ ] Fixes an issue where images had a 0 width or height throwing an error
* [ ] Fixes and issue with underlines
* [ ] Fixes and issue with font loading
* [ ] improves cleaning up of frames when running many invocations async
* [ ] fixes some cors and browser extension image handling
* [ ] improves performance when running many invocations async on the same page by using a global cache

Explain the **motivation** for making this change. What existing problem does the pull request solve?

**Test plan (required)**

Demonstrate how the issue/feature can be replicated. For most cases, simply adding an appropriate html/css template into the [reftests](https://github.com/html2canvas/html2canvas/tree/master/tests/reftests) should be sufficient. Please see other tests there for reference.

**Code formatting**

Please make sure that code adheres to the project code formatting. Running `npm run format` will automatically format your code correctly.

**Closing issues**

Fixes #

### CODE REVIEW

**Comment by Zefling** [comment]
Why keep the `_` for `_cache`? It's a public value.

**Comment by jamie-zaharopoulos-fh** [comment]
> Why keep the `_` for `_cache`? It's a public value.

Just a carry over from existing naming. fixed it in the last commit. Any other PR comments/suggestions?

File: src/dom/node-parser.ts [hunk 1]
```diff
@@ -19,6 +19,26 @@ const parseNodeTree = (context: Context, node: Node, parent: ElementContainer, r
         nextNode = childNode.nextSibling;
 
         if (isTextNode(childNode) && childNode.data.trim().length > 0) {
+            // The U tag marks text with a special underline treatment, and it's not possible to get the underline style from the browser's computed style.
+            const parentStep = 3;
+            let hasUnderline,
+                // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                pNode: any = childNode;
```
Comment (Zefling) [inline src/dom/node-parser.ts:26]: not use `,` separator ;)

File: src/dom/node-parser.ts [hunk 2]
```diff
@@ -19,6 +19,26 @@ const parseNodeTree = (context: Context, node: Node, parent: ElementContainer, r
...
+                    hasUnderline = true;
+                    break;
+                }
+                pNode = pNode.parentElement;
+            }
+            if (hasUnderline && parent.styles.textDecorationLine) {
+                for (let j = 0; j < parent.styles.textDecorationLine.length; j++) {
+                    parent.styles.textDecorationLine[j] = 1;
+                }
+            }
```
Comment (Zefling) [inline src/dom/node-parser.ts:41]: Probably better
```
        const line = parent.styles.textDecorationLine;
            if (hasUnderline && line) {
                for (let j = 0; j < line.length; j++) {
                    line[j] = 1;
                }
            }
````

File: src/render/canvas/canvas-renderer.ts [hunk 3]
```diff
@@ -327,7 +327,15 @@ export class CanvasRenderer extends Renderer {
                 await container.setup(image);
                 this.renderReplacedElement(container, curves, image);
             } catch (e) {
-                this.context.logger.error(`Error loading image ${container.src}`);
+                try {
+                    if (this.context.cache.deleteImage(container.src) && e.type && e.type === 'error') {
```
Comment (Zefling) [inline src/render/canvas/canvas-renderer.ts:331]: `e.type && e.type === 'error'`

Just `e.type === 'error'` for same effect

File: src/dom/node-parser.ts [hunk 4]
```diff
@@ -19,6 +19,26 @@ const parseNodeTree = (context: Context, node: Node, parent: ElementContainer, r
...
+                // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                pNode: any = childNode;
+            for (let i = 0; i < parentStep; i++) {
+                if (!pNode) {
+                    break;
+                }
+                if (pNode.parentElement && pNode.parentElement.tagName && pNode.parentElement.tagName === 'U') {
+                    hasUnderline = true;
+                    break;
+                }
```
Comment (Zefling) [inline src/dom/node-parser.ts:34]: `if (pNode.parentElement?.tagName === 'U')`

for same effect

**Pull request merged**